index.rst

```diff
@@ -4,9 +4,164 @@ Writeable Butlers for Science Users
+This may actually be faster for ``DirectButler`` as well, since we often want to join in data IDs that are a small subset of those in the temporary table (and a subset that will have already been computed in the client).
+If that turns out to be the case, we should probably drop temporary table support from the query system, as ``QuantumGraph`` generation was the driving user case for it.
+We should probably drop it from ``RemoteButler`` regardless, as emulated support is probably worse than no support.
+
+QuantumBackedButler Transfers from RemoteButler to DirectButler
+---------------------------------------------------------------
+
+After a ``QuantumGraph`` is generated against the official data repository, we need to be able to execute that graph such that:
+
+- input datasets are fetched from the ``RemoteButler`` to the offical data repository (with URL signing as needed);
+- output datasets are written into the datastore root of the personal data repository;
+- the "transfer" stage that inserts rows into the personal data repository's database can (optionally) transfer inputs as well as outputs, including setting up the collection structure in which the inputs were found in the offical data repository.
+
+This requires additional persistent information that would have to be put into the ``QuantumGraph`` or some other file.
+
+Signed URLs in DirectButler
+---------------------------
+
+In order for a personal data repository to reference an official data product (instead of copying it), the ``DirectButler`` client will need to be able to obtain signed URLs from the ``RemoteButler`` server.
+This could be implemented as a new ``Datastore`` subclass (to be chained in with a regular ``FileDatastore`` for the personal datasets in that repository), an extension to the current ``FileDatastore``.
+It could also be implemented directly within ``DirectButler``, but since we want this functionality in ``QuantumBackedButler`` as well, implementing it in ``Datastore`` seems to make more sense.
+
+User Interfaces
+---------------
+
+We need high-level command-line and notebook-friendly tools for at least the following tasks:
+
+- Create a new, empty satellite data repository from the official data repository.
+  This needs to set up a datastore that can reference official datasets (with URL signing), and it may be convenient for it to always register a few instruments and/or skymaps, and maybe even write curated calibrations.
+
+- Create and run a ``QuantumGraph`` with inputs from the official data repository or the personal data repository and outputs going to the personal data repository, using ``QuantumBackedButler``.
+  In the minimum-viable system the input source would strictly be one repository OR the other, but in the future we may be able to support combinations, and should plan for that in the UI.
```

Comment:
Does this document need to address questions like the following regarding the situation when "take from both" is an option:

* Is the personal data repository always "ahead" of the upstream "official" one in any search?
* Can chained collections be created in the person repo across collections in both repos?  (Obviously related to the above.)  Tagged collections?
* Does this model seem to support cases where a user wants to create new calibrations and then apply them to the "official" data?

Reply:
Supporting general multi-repo butler queries is a lot harder than multi-repo QG generation, which I envision working by submitting different queries to each butler and joining the results in-memory.  There would be no cross-repo collections; if the users wants a CHAINED collection with both inputs and outputs, we'd put references to the official-repo datasets into the user registry.

Maintaining that sort of mirror input collection (e.g. if the user then makes another QG that wants more inputs) is one of the big pieces of complexity in this scheme.